ENH: support additional dtypes in pad_nd

shubham-61969 opened the pull request:
Prefer the PyTorch padding backend when supported and safely fall back
to NumPy on error. Add unit tests to validate backend selection and
ensure output dtype is preserved.

Fixes #7842

### Description

This pull request relaxes dtype restrictions in `pad_nd` and prefers
the PyTorch padding backend when supported, with a safe fallback to
NumPy on error. This enables support for additional dtypes (e.g. bool)
that are already handled correctly by recent PyTorch versions.

Unit tests are added to validate backend selection and ensure dtype
preservation.

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.

shubham-61969 committed:
ENH: support additional dtypes in pad_nd

Prefer the PyTorch padding backend when supported and safely fall back
to NumPy on error. Add unit tests to validate backend selection and
ensure output dtype is preserved.

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 committed:
FIX: ensure pad_nd handles value kwarg correctly across modes

Strip value for non-constant modes for both PyTorch and NumPy backends
and ensure reliable fallback behavior. Update tests to cover multiple
padding modes and dtype preservation.

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 committed:
FIX: clean pad_nd fallback handling and mode-specific kwargs in test

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ericspod commented on monai/transforms/croppad/functional.py:
```diff
@@ -96,19 +96,24 @@ def pad_nd(
-        return _pad(img, pad_width=to_pad, mode=mode, **kwargs)
+        call_kwargs = dict(kwargs)
+        if mode != "constant":
+            call_kwargs.pop("value", None)
```
If this condition isn't present, the effect of providing a `value` argument and not use "constant" mode is to raise an exception in the pad routine used, I think this is intended behaviour. Here if the `value` argument is removed this silently allows unintended arguments to be ignored, it's better to raise an exception instead.

ericspod commented on tests/transforms/croppad/test_pad_nd_dtypes.py:
```diff
@@ -0,0 +1,70 @@
+# Copyright (c) MONAI Consortium
```
In MONAI we use `unittest` and `parameterized` packages for tests and not `pytest`, specifically we use test classes and methods for unit tests rather than functions. Please reformulate these tests to use these packages according the style of other tests that are present here. It might make sense to add your tests to an existing file rather than a new one, have a look at existing files to see if it does make sense that way.

ericspod reviewed (commented):
Hi @shubham-61969 thanks for the contribution. I think the change itself is fine with a few comments, the tests do need to be reformulated with `unittest` in particular. I think the previous implementation with its limitations was a result of older PyTorch versions so it's good to get this fix in. Please have a look again and then we can rereview.

shubham-61969 committed:
FIX: align pad_nd behavior and tests with MONAI conventions

- Raise an explicit error when alue is provided with non-constant modes.
- Rewrite tests using unittest + parameterized to match MONAI style.

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 committed:
Resolve merge conflict in pad_nd tests

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

shubham-61969 committed:
Resolve incorrectly Imports sorted and/or formatted.

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 committed:
Merge branch '7842-pad-nd-more-dtypes' of https://github.com/shubham-61969/MONAI into 7842-pad-nd-more-dtypes

shubham-61969 commented:
Hi @ericspod , thanks for the feedback.

I’ve updated the implementation to explicitly raise an error when value is provided with modes other than "constant", so the original contract is preserved. I also reworked the tests to use unittest + parameterized and aligned them with MONAI’s existing test style.

All GitHub CI checks are now passing.

Could you please take another look? If there are any further changes you’d like, I’m happy to address them otherwise I believe this should be ready to merge.

ericspod commented on monai/transforms/croppad/functional.py:
```diff
@@ -96,19 +96,22 @@ def pad_nd(
+            # Try PyTorch pad for these modes; fallback to NumPy on error.
             _pad = _pt_pad
-        return _pad(img, pad_width=to_pad, mode=mode, **kwargs)
+        call_kwargs = dict(kwargs)
```
We don't actually need `call_kwargs` now as you can refer to `kwargs` directly.

shubham-61969 replied:
My bad, I completely missed switching it back to kwargs during the refactor. Thanks for pointing that out!

ericspod commented on tests/transforms/croppad/test_pad_nd_dtypes.py:
```diff
@@ -0,0 +1,101 @@
+        self.assertEqual(out.dtype, img.dtype)
+        self.assertEqual(out.shape, (1, 6, 8))
+
+    @parameterized.expand([torch.bool, torch.int8, torch.int16, torch.int32, torch.int64, torch.uint8, torch.float32])
```
Often it's cleaner to put arguments to `expand` into a variable at the top of the file, here and in the next usage it just is cleaner to read the tests.

shubham-61969 replied:
Noted!

ericspod commented on tests/transforms/croppad/test_pad_nd_dtypes.py:
```diff
@@ -0,0 +1,101 @@
+        img = torch.ones((1, 4, 4))
+        to_pad = [(0, 0), (1, 1), (2, 2)]
+        with self.assertRaises(ValueError):
+            pad_nd(img, to_pad, mode="reflect", **{"value": 0})
```
```suggestion
            pad_nd(img, to_pad, mode="reflect", value=0)
```

ericspod reviewed (approved):
Looks good to me now, I had some minor comments only. We'll try to get this merge shortly when other things have been resolved.

shubham-61969 committed:
address minor changes review comments in pad_nd and tests

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 committed:
FIX: apply CodeRabbit suggestions for pad_nd value handling

Signed-off-by: Shubham Chandravanshi <shubham.chandravanshi378@gmail.com>

shubham-61969 commented:
Hello @ericspod, 
All suggested changes are now applied would appreciate a quick re-review when you have a moment.
Thanks!

ericspod reviewed (approved):
Hi @shubham-61969 sorry for the long delay, I think this looks good now, thanks!

ericspod committed:
Merge branch 'dev' into 7842-pad-nd-more-dtypes
